Consistent DrawPoint and DrawCheckMark on all platforms

MaartenBent opened the pull request:
Use the generic `DrawCheckMark` implementation in MSW `wxDC`, so it looks the same across all platforms and renderers and not like a Windows control.

Before - After
![before](https://user-images.githubusercontent.com/8088070/62651680-da777300-b959-11e9-9a42-28e8de26c66d.png)![after](https://user-images.githubusercontent.com/8088070/62651681-db100980-b959-11e9-90e9-92ac94685447.png)

Fix `DrawPoint` in Direct2D renderer.
Draw a point with the `wxPen` width in MSW `wxDC`, similar to other platforms and renderers.

Update the drawing sample to simplify switching between renderers, and add an option the toggle anti-aliasing.

MaartenBent committed:
Use generic DrawCheckMark in MSW wxDC

Use the generic implementation, so it looks like the check marks on all other
platforms and graphics contexts and not like a Windows control.

MaartenBent committed:
Use numbers for accelerator keys to select renderer in drawing sample

Determine if renderer is selected by comparing the renderer itself, not by name.

MaartenBent committed:
Add option to drawing sample to toggle anti-aliasing of graphics renderer

vadz commented on samples/drawing/drawing.cpp:
```diff
@@ -2142,17 +2157,17 @@ MyFrame::MyFrame(const wxString& title, const wxPoint& pos, const wxSize& size)
-    menuFile->AppendCheckItem(File_GC_Default, "Use default wx&GraphicContext\tCtrl-Y");
-    m_menuItemUseDC = menuFile->AppendRadioItem(File_DC, "Use wx&DC\tShift-Ctrl-Y");
+    menuFile->AppendCheckItem(File_GC_Default, "Use default wx&GraphicContext\tCtrl-1");
+    m_menuItemUseDC = menuFile->AppendRadioItem(File_DC, "Use wx&DC\tCtrl-2");
```
Really a nitpick, but should this be `Ctrl-0` perhaps? Using DC is different from all the other ones...

MaartenBent replied:
Done, also removed `Ctrl`.

vadz commented on src/common/dcgraph.cpp:
```diff
@@ -752,9 +752,16 @@ void wxGCDCImpl::DoDrawPoint( wxCoord x, wxCoord y )
-    if (m_graphicContext->GetRenderer() == wxGraphicsRenderer::GetGDIPlusRenderer())
+#if defined(__WXMSW__)
+    bool useStrokeLine = false;
+#if wxUSE_GRAPHICS_GDIPLUS
```
At this point (pun not intended), I'd rather add a `virtual CanStrokeSinglePointLine()` or something like this to `wxGraphicsRenderer`.

MaartenBent replied:
This still wouldn't cover the difference between macOS and Cairo.
Maybe I should add `wxGraphicsContext::DrawPoint()`?

vadz replied:
Yes, this is probably an even better idea.

MaartenBent replied:
I found an implementation (using `AddCircle`) that works for all renderers.

vadz reviewed (approved):
It's a bit bothersome that this is going to silently change the appearance of programs using `DrawCheckMark()`, but I guess we can afford this because `wxDC` is really supposed to result in the same output under all platforms, but ideal would be to at least add `wxRendererNative` method that could be used to draw a native-looking one check mark.

Also, if we decide that this should be the same for all platforms, shouldn't we get rid of `DoDrawCheckMark()` entirely and just implement `DrawCheckMark()` in common code (the same applies to `DoCrossHair()` as well BTW)?

Just to be clear, I'm fine with applying this in its current state, but if any of the proposed enhancements make sense to you, please let me know if you'd prefer to wait until you update this PR with them.

Thanks!

MaartenBent commented:
I can add `DrawCheckMark` to wxRendererXP/wxRendererMSW/wxRendererGeneric/wxRendererNative that will use the original checkmark and fall back to `wxDC::DrawCheckMark` on other platforms.

`wxDCMirror` has its own implementation of `DoDrawCheckMark` so I don't think it can be removed.

I'll have a look at `DoCrossHair` later and see if it can be generalized for all platforms. but not in this PR.

MaartenBent committed:
Add DrawCheckMark and GetCheckMarkSize to wxRendererNative

Show its use in the render sample. Also use the recently added GetExpanderSize
for the size of DrawTreeItemButton.

paulcor commented on src/common/dcgraph.cpp:
```diff
@@ -752,9 +752,16 @@ void wxGCDCImpl::DoDrawPoint( wxCoord x, wxCoord y )
+#if wxUSE_GRAPHICS_DIRECT2D
+    useStrokeLine |= m_graphicContext->GetRenderer() == wxGraphicsRenderer::GetDirect2DRenderer();
+#endif
+    // single point path does not work with GDI+ and D2D
```
I'm pretty sure single point path does work with D2D. What exactly is wrong with it?

MaartenBent replied:
See below screenshots of the drawing sample were I added the following code to `DrawDefault` and changed the background colour to black:
```c++
    dc.SetPen(wxPenInfo(*wxGREEN, 10, wxPENSTYLE_SOLID));
    dc.DrawPoint(25, 25);
    dc.DrawPoint(50, 25);
```

You can see D2D draws nothing, and msw wxDC does not use the pen width.

I tested again and you're right that D2D does not need the GDI+ work-around (adding dx/dy), it can use `StrokeLine(x, y, x, y)` directly. The current implementation using `wxGraphicsPath` does not work.

msw wxDC - GDI+ - D2D
<a href="https://user-images.githubusercontent.com/8088070/62682741-2c9db000-b9bd-11e9-8bb1-72d18ee17fd5.png"><img src="https://user-images.githubusercontent.com/8088070/62682741-2c9db000-b9bd-11e9-8bb1-72d18ee17fd5.png" width="150" alt="msw wxDC"/></a><a href="https://user-images.githubusercontent.com/8088070/62682745-30313700-b9bd-11e9-9786-0611916a7c30.png"><img src="https://user-images.githubusercontent.com/8088070/62682745-30313700-b9bd-11e9-9786-0611916a7c30.png" width="150" alt="msw GDI+"/></a><a href="https://user-images.githubusercontent.com/8088070/62682748-31fafa80-b9bd-11e9-90f0-28a6b0ef08fe.png"><img src="https://user-images.githubusercontent.com/8088070/62682748-31fafa80-b9bd-11e9-90f0-28a6b0ef08fe.png" width="150" alt="msw D2D"/></a>

Ubuntu x11(?) wxDC - cairo - macOS core graphics
<a href="https://user-images.githubusercontent.com/8088070/62682762-36bfae80-b9bd-11e9-87d3-779bac44544b.png"><img src="https://user-images.githubusercontent.com/8088070/62682762-36bfae80-b9bd-11e9-87d3-779bac44544b.png" width="150" alt="gtk wxDC"/></a><a href="https://user-images.githubusercontent.com/8088070/62682765-38897200-b9bd-11e9-95bd-65c4197f5c8d.png"><img src="https://user-images.githubusercontent.com/8088070/62682765-38897200-b9bd-11e9-95bd-65c4197f5c8d.png" width="150" alt="gtk ciaro"/></a><a href="https://user-images.githubusercontent.com/8088070/62682770-3aebcc00-b9bd-11e9-81c2-457af01127c4.png"><img src="https://user-images.githubusercontent.com/8088070/62682770-3aebcc00-b9bd-11e9-81c2-457af01127c4.png" width="150" alt="macOS gc"/></a>

paulcor replied:
I'm sure I tested all the `wxGraphicsContext` backends when I added `wxGCDCImpl::DoDrawPoint()`. I will investigate.

It looks like you tested with GTK3, where `wxDC` uses Cairo. If you try it with GTK2, I think you'll see that it behaves the same as MSW. Which is as described in the current documentation. My take on `DrawPoint()` is that it works properly with unit scale and pen width, anything else may or may not work.

MaartenBent replied:
> It looks like you tested with GTK3

You're right. I updated the screenshot with GTK2 version. It's the same as wxDC on MSW.

What would you propose (besides fixed D2D)? Update `wxDC` to draw bigger points with penwidth >1. Or always draw a point of 1 pixel in `wxGraphicsContext`. Or just document the difference of DrawPoint between `wxDC` and `wxGCDC`?

paulcor replied:
`DrawPoint()` did work with D2D when I checked it in, but was subsequently broken by c8fe811636

a-wi replied:
This regression with D2D sub-path should be fixed in 283ac3096ddd431b19b36eeb903ba9ed31bcb674.

MaartenBent replied:
Thanks. I'll revert my commit.

MaartenBent commented:
I updated the `DrawPoint` implementation of `wxGCDC` so the same one can be used for all renderers.
Currently it does use the pen width, so there should be no noticeable change with the previous implementation.

MaartenBent committed:
Use default wxPen for DrawPoint in wxGCDC

Match the default wxDC implementations by using a default wxPen with width 1,
wxPENSTYLE_SOLID and default join and cap values.

vadz commented:
Thanks again!
